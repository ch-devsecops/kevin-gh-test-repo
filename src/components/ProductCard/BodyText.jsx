import React, { useContext, useLayoutEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { Markdown, Box } from '@honda-canada/design-system-react';
import { isEditorActive } from '@sitecore-jss/sitecore-jss-react';
import isSSR from '../../utils/isSSR';
import SimilarRecentlyViewedContext from '../SimilarRecentlyViewedProducts/SimilarRecentlyViewedContext';
import SimpleStyle from '../../utils/propTypes/simple';

const BodyText = ({ text, py, mt, size }) => {
  const { minBodyTextHeight, setMinBodyTextHeight } = useContext(SimilarRecentlyViewedContext);

  const bodyTextRef = useRef();

  useLayoutEffect(() => {
    if (isEditorActive() || isSSR()) return;
    // dynamically adjust bodyText height to maintain cards alignment
    const measuredHeight = parseInt(
      getComputedStyle(bodyTextRef.current).getPropertyValue('height').replace('px', ''),
      10,
    );

    if (measuredHeight > minBodyTextHeight) {
      setMinBodyTextHeight(measuredHeight);
    }
  }, [minBodyTextHeight]);

  return (
    <Box
      ref={bodyTextRef}
      flexGrow="1 !important"
      minHeight={['default', 'default', minBodyTextHeight]}
      py={py}
      mt={mt}
    >
      <Markdown size={size}>{text}</Markdown>
    </Box>
  );
};

BodyText.defaultProps = {
  size: ['xSmall', 'small'],
};

BodyText.propTypes = {
  text: PropTypes.string,
  size: SimpleStyle,
};

export default BodyText;
