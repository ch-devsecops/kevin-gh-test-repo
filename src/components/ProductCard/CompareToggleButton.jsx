import React, { useContext, useState, useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';
import { Box, Label } from '@honda-canada/design-system-react';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import themeStyles from './ProductCard.styles';
import LocalStorageContext from '../LocalStorageContext/LocalStorageContext';
import { compareProducts } from '../../utils/constants';
import { ADDED_COMPARISON_ITEM, REMOVED_COMPARISON_ITEM } from './utils';

const Container = themeStyles.apply(Box, 'Container');
const CompareLabel = themeStyles.apply(Label, 'CompareLabel');
const CheckboxInput = styled('input').attrs({ type: 'checkbox', tabIndex: '-1', autoComplete: 'off' })(
  themeStyles.get('CheckboxInput'),
);
const ToggleWrapper = styled('label')(themeStyles.get('ToggleWrapper'));
const StyledCheckmark = styled('span')(
  ({ disabled, checked, theme }) => `
  position: absolute;
  top: 0;
  left: 0;
  height: 20px;
  width: 20px;

  &::before {
    content: '';
    position: absolute;
    display: block;
    background-color: ${disabled ? theme.colors.grey[0] : theme.colors.red};
    transition: 0.4s all;
    bottom: 4px;
    left: 35%;
    width: 5%;
    height: 55%;
    border: solid ${disabled ? theme.colors.grey[0] : theme.colors.red};
    border-width: 0.1px;
    transition: 0.4s all;
    -ms-transform: rotate(${checked ? '45deg' : '180deg'});
    transform: rotate(${checked ? '45deg' : '180deg'});
  }

  &::after {
    content: '';
    position: absolute;
    display: block;
    background-color: ${disabled ? theme.colors.grey[0] : theme.colors.red};
    transition: 0.4s all;
    bottom: 4px;
    left: 35%;
    width: 5%;
    height: 55%;
    border: solid ${disabled ? theme.colors.grey[0] : theme.colors.red};
    border-width: 0.1px;
    transition: 0.4s all;
    -ms-transform: rotate(${checked ? '135deg' : '90deg'});
    transform: rotate(${checked ? '135deg' : '90deg'});
  }

  ${CheckboxInput}:disabled + & {
    border-color: ${theme.colors.grey[1]};
  }
`,
);

const CompareToggleButton = ({
  id,
  ariaLabel,
  compareLabelFontSize,
  compareLabelLineHeight,
  gtmTags,
  detId,
  ...rest
}) => {
  const { toCompareProducts, setToCompareProducts, removeFromCompareProducts } = useContext(LocalStorageContext);
  const [gtmTitle, setGtmTitle] = useState(ADDED_COMPARISON_ITEM);
  const { t } = useTranslation();

  const isCompared = useMemo(() => toCompareProducts.includes(detId), [toCompareProducts, detId]);

  useEffect(() => {
    if (isCompared) {
      setGtmTitle(REMOVED_COMPARISON_ITEM);
    } else {
      setGtmTitle(ADDED_COMPARISON_ITEM);
    }
  }, [isCompared]);

  const onCompareChange = () => {
    if (isCompared) {
      removeFromCompareProducts(detId);
    } else {
      setToCompareProducts(detId);
    }
  };

  const compareIconLabel = isCompared
    ? t('Shared.Common.removeFromCompareButton')
    : t('Shared.Common.addToCompareButton');

  const isDisabled = toCompareProducts.length >= compareProducts.limit && !isCompared ? 'disabled' : undefined;

  const stopPropagationHandler = event => {
    event.stopPropagation();
  };

  return (
    <Container mt="0" onClick={stopPropagationHandler}>
      <ToggleWrapper
        data-testid="product-card-compare-btn"
        data-testvalue={!!isCompared}
        {...gtmTags}
        data-gtm-title={gtmTitle}
      >
        <CheckboxInput
          id={id}
          containerSize="small"
          checked={isCompared}
          disabled={isDisabled}
          onChange={onCompareChange}
          aria-checked={isCompared}
          aria-disabled={isDisabled}
          aria-label={ariaLabel}
          {...gtmTags}
          data-gtm-title={gtmTitle}
          {...rest}
        />
        <StyledCheckmark disabled={isDisabled} checked={isCompared} data-testid="cy-compare-toggle" />
        <CompareLabel
          disabled={isDisabled}
          htmlFor={id}
          compareLabelFontSize={compareLabelFontSize}
          compareLabelLineHeight={compareLabelLineHeight}
        >
          {compareIconLabel}
        </CompareLabel>
      </ToggleWrapper>
    </Container>
  );
};
CompareToggleButton.defaultProps = {
  compareLabelFontSize: '14px',
  compareLabelLineHeight: '16px',
};

CompareToggleButton.propTypes = {
  id: PropTypes.string,
  ariaLabel: PropTypes.string,
  compareLabelFontSize: PropTypes.string,
  compareLabelLineHeight: PropTypes.string,
  gtmTags: PropTypes.shape({
    'data-gtm-interaction-type': PropTypes.string,
    'data-gtm-category': PropTypes.string,
    'data-gtm-component-type': PropTypes.string,
    'data-gtm-body-style': PropTypes.string,
    'data-gtm-model': PropTypes.string,
  }),
  detId: PropTypes.string.isRequired,
};

export default CompareToggleButton;
