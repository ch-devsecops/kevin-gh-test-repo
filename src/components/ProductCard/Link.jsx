import React from 'react';
import { Link as DsrLink } from '@honda-canada/design-system-react';
import PropTypes from 'prop-types';

const Link = ({ href, onClick, gtmTags, children, ...rest }) => (
  <DsrLink as="a" disableHover href={href} onClick={onClick} {...gtmTags} {...rest}>
    {children}
  </DsrLink>
);

Link.propTypes = {
  href: PropTypes.string,
  onClick: PropTypes.func,
  gtmTags: PropTypes.shape({
    'data-gtm-interaction-type': PropTypes.string,
    'data-gtm-component-type': PropTypes.string,
    'data-gtm-body-style': PropTypes.string,
    'data-gtm-model': PropTypes.string,
    title: PropTypes.string,
  }),
  children: PropTypes.shape({}),
  rest: PropTypes.shape({}),
};
export default Link;
