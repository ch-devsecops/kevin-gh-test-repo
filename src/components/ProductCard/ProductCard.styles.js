import css from '@styled-system/css';
import { ThemeStyles } from '@honda-canada/design-system-react';

const themeStyles = new ThemeStyles();

themeStyles.addBaseStyles('StyledCopy')(({ theme }) => {
  css({
    fontFamily: theme.fonts.default,
  });
});

themeStyles.addBaseStyles('Container')(() =>
  css({
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    pt: 's',
    pb: ['s', 'l'],
  }),
);

themeStyles.addBaseStyles('Wrapper')(({ theme, maxWidth = ['246px', '295px'] }) =>
  css({
    display: 'flex',
    flexDirection: 'column',
    flexGrow: 1,
    fontFamily: theme.fonts.default,
    textAlign: 'center',
    maxWidth,
    p: ['xs', 'default'],
    mb: 0,
    '&:hover': {
      backgroundColor: ['none', 'grey.5'],
    },
  }),
);

themeStyles.addBaseStyles('SecondaryButton')(({ display = 'block' }) =>
  css({
    display,
    mt: 'default',
    mb: 's',
  }),
);

themeStyles.addBaseStyles('ToggleWrapper')(() =>
  css({
    display: 'flex',
    alignItems: 'center',
    gap: 's',
    position: 'relative',
    cursor: 'pointer',
    height: '20px',
  }),
);

themeStyles.addBaseStyles('CompareLabel')(({ disabled, compareLabelFontSize, compareLabelLineHeight }) =>
  css({
    fontFamily: 'bold',
    ml: 'xs',
    fontSize: compareLabelFontSize,
    lineHeight: compareLabelLineHeight,
    pointerEvents: 'none',
    color: disabled ? 'grey.0' : undefined,
  }),
);

themeStyles.addBaseStyles('CheckboxInput')(() =>
  css({
    height: 0,
    width: 0,
    opacity: 0,
    zIndex: -1,
  }),
);

themeStyles.addBaseStyles('CashSavings')(() =>
  css({
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    textAlign: 'center',
  }),
);

export default themeStyles;
