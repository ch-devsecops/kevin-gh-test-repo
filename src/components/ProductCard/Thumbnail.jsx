import React from 'react';
import PropTypes from 'prop-types';
import { Box, Image } from '@honda-canada/design-system-react';

const Thumbnail = ({ image, ...styleProps }) => {
  const { src, alt } = image;
  return (
    <Box height={['139px', '248px']} {...styleProps}>
      <Image style={{ objectFit: 'contain' }} src={src} alt={alt} />
    </Box>
  );
};

Thumbnail.propTypes = {
  image: PropTypes.shape({
    src: PropTypes.string,
    alt: PropTypes.string,
  }),
};

export default Thumbnail;
