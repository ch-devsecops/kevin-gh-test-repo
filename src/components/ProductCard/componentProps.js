import PropTypes from 'prop-types';

const compnentProps = {
  pricing: PropTypes.bool,
  fields: PropTypes.shape({
    modelName: PropTypes.shape({
      value: PropTypes.string,
    }),
    tagline: PropTypes.shape({
      value: PropTypes.string,
    }),
    primaryImage: PropTypes.shape({
      value: PropTypes.shape({
        images: PropTypes.arrayOf(
          PropTypes.shape({
            url: PropTypes.string,
            alt: PropTypes.shape({
              value: PropTypes.string,
            }),
          }),
        ),
      }),
    }),
    thumbnail: PropTypes.shape({
      value: PropTypes.shape({
        images: PropTypes.arrayOf(
          PropTypes.shape({
            url: PropTypes.string,
            alt: PropTypes.shape({
              value: PropTypes.string,
            }),
          }),
        ),
      }),
    }),
    modelPage: PropTypes.shape({
      field: PropTypes.shape({
        fields: PropTypes.shape({
          url: PropTypes.string,
        }),
      }),
    }),
    id: PropTypes.string,
    name: PropTypes.string,
  }),
  gtmTags: PropTypes.shape({
    title: PropTypes.string,
    compareTitle: PropTypes.string,
    seriesName: PropTypes.string,
    componentName: PropTypes.string,
    crankshaftName: PropTypes.string,
    interactionType: PropTypes.string,
    compareInteractionType: PropTypes.string,
  }),
};

export default compnentProps;
