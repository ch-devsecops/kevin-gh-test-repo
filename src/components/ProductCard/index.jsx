import React from 'react';

import ProductCardUI from './ProductCard';
import { gtmEngineMap, gtmMarineMap, variant1, variant2, variant3 } from './utils';
import { ENGINE_SITE_NAME, MARINE_SITE_NAME } from '../../utils/constants';
import useAppName from '../../utils/sitecoreContext/useAppName';

const ProductCardMap = props => {
  const appName = useAppName();
  const SIMILAR_RECENTLY_VIEWED_COMPONENT_NAME = 'SimilarRecentlyViewedProducts';

  const { fields, gtmTags } = props || {};

  let gtmProps;
  let variant;
  switch (appName) {
    case MARINE_SITE_NAME:
      gtmProps = gtmMarineMap(fields, gtmTags);
      variant = gtmTags?.componentName === SIMILAR_RECENTLY_VIEWED_COMPONENT_NAME ? variant3 : variant2;
      break;

    case ENGINE_SITE_NAME:
    default:
      gtmProps = gtmEngineMap(fields, gtmTags);
      variant = variant1;
      break;
  }

  return <ProductCardUI {...props} variant={variant} {...gtmProps} />;
};

export default ProductCardMap;
