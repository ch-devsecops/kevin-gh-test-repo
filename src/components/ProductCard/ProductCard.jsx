import React, { useContext } from 'react';
import { Box, Button, Optional } from '@honda-canada/design-system-react';
import { useTranslation } from 'react-i18next';
import Link from './Link';
import { wrapJSSFields } from '../../utils/wrapJSSFields';
import CompareToggleButton from './CompareToggleButton';
import BodyText from './BodyText';
import Title from './Title';
import Thumbnail from './Thumbnail';
import themeStyles from './ProductCard.styles';
import LocalStorageContext from '../LocalStorageContext/LocalStorageContext';
import componentProps from '../MarineProductDetailsCard/componentProps';
import { useConfiguration, VIEW_DETAILS } from './utils';
import PriceComponent from '../PriceComponent';
import SaveLabel from '../PriceComponent/SaveLabel';
import { EqualHeightElement } from '../../utils/components/EqualHeight';

const Wrapper = themeStyles.apply(Box, 'Wrapper');
const SecondaryButton = themeStyles.apply(Button, 'SecondaryButton');
const CashSavingsContainer = themeStyles.apply(Box, 'CashSavings');

const ProductCard = ({ fields, gtmTags, variant, prices, isFetching, hasError, ...styledProps }) => {
  const { t } = useTranslation();

  const { setRecentlyViewedProducts } = useContext(LocalStorageContext);

  const {
    imageOrder,
    showPricing,
    bodyMarginTop,
    bodySize,
    wrapperMaxWidth,
    ctaButtonDisplay,
    priceComponentStyles,
    compareLabelFontSize,
    compareLabelLineHeight,
  } = useConfiguration(variant);

  const { detIdentifier, tagline, thumbnail, modelPage, id, primaryImage } = wrapJSSFields(fields);
  const modelPageUrl = modelPage?.field?.fields?.url;

  const image = primaryImage
    ? {
        src: primaryImage.field?.item?.value?.src,
        alt: primaryImage.field?.item?.value?.alt,
      }
    : {
        src: thumbnail?.value?.images?.[0]?.url,
        alt: thumbnail?.value?.images?.[0]?.alt?.value,
      };
  const productId = id?.field;
  const detId = detIdentifier?.value;

  let hasDiscount;
  let discountValue;

  const saveStyles = {
    copy: {
      styles: {
        fontSize: ['14px', '14px', '14px'],
      },
    },
  };

  if (prices) {
    discountValue = prices?.discount?.priceDiscountAmount;
    hasDiscount = !!discountValue;
  }

  const handleLinkOnClick = () => setRecentlyViewedProducts(detId);
  return (
    <Wrapper maxWidth={wrapperMaxWidth} {...styledProps} data-testid="product-card">
      <EqualHeightElement placeholder={!hasDiscount}>
        <Optional when={hasDiscount}>
          <CashSavingsContainer>
            <SaveLabel value={discountValue} saveStyles={saveStyles} />
          </CashSavingsContainer>
        </Optional>
      </EqualHeightElement>
      <EqualHeightElement>
        <Link href={modelPageUrl} onClick={handleLinkOnClick} gtmTags={gtmTags?.link}>
          <Title mt="0" title={fields?.name} />
        </Link>
      </EqualHeightElement>

      <EqualHeightElement placeholder={!showPricing} inherit>
        <Optional when={showPricing}>
          <PriceComponent
            prices={prices}
            priceComponentStyles={priceComponentStyles}
            isFetching={isFetching}
            hasError={hasError}
          />
        </Optional>
      </EqualHeightElement>
      <Link href={modelPageUrl} onClick={handleLinkOnClick} gtmTags={gtmTags?.link}>
        <Thumbnail order={imageOrder} image={image} />
      </Link>
      <BodyText text={tagline?.value} mt={bodyMarginTop} size={bodySize} />
      <SecondaryButton
        tabIndex={0}
        as="a"
        display={ctaButtonDisplay}
        href={modelPageUrl}
        styling="secondary"
        onClick={() => handleLinkOnClick(VIEW_DETAILS)}
        {...gtmTags?.cta}
      >
        {t('Shared.Common.viewDetailsButton')}
      </SecondaryButton>
      <CompareToggleButton
        compareLabelFontSize={compareLabelFontSize}
        compareLabelLineHeight={compareLabelLineHeight}
        tabIndex={0}
        id={productId}
        ariaLabel={t('Shared.Common.compareCheckboxAria', { modelName: fields?.name })}
        gtmTags={gtmTags?.compare}
        detId={detId}
      />
    </Wrapper>
  );
};

ProductCard.propTypes = componentProps;

export default ProductCard;
