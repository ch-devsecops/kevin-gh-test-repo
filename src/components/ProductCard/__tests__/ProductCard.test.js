import React from 'react';
import { jssRender, cleanup, axe } from 'test-utils';
import ProductCard from '..';

afterEach(cleanup);

const mockData = require('../../../../sample-mock-data/routes/product-card/en.json');

describe('<ProductCard>', () => {
  test('should render correctly', () => {
    const { container } = jssRender(<ProductCard fields={mockData} />);

    expect(container).toMatchSnapshot();
    expect(container.children[0]).toBeTruthy();
  });

  test('should have no accessibility violations', async () => {
    const { container } = jssRender(<ProductCard fields={mockData} />);

    expect(await axe(container)).toHaveNoViolations();
  });
});
