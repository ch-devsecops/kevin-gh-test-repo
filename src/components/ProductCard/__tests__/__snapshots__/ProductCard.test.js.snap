// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard> should render correctly 1`] = `
<div>
  <div
    class="Box-sc-1bsf784-0 sc-1k44dfv-0 ljGqou"
    data-testid="product-card"
  >
    <div
      style="overflow: hidden; transition-property: height; display: block; transition-duration: 0.25s;"
    />
    <div
      style="overflow: hidden; transition-property: height; display: block; transition-duration: 0.25s;"
    >
      <a
        class="Link-sc-16wd4a0-0 bXBGRT"
        data-gtm-interaction-type="explore"
        data-gtm-trim="GX100"
        title="view details"
      >
        <div
          class="sc-1k44dfv-0 cALsOM"
          mb="xs,s"
          mt="0"
        >
          

          <h5
            class="sc-1k44dfv-0 gXWcpP"
          >
            GX100
          </h5>
        </div>
      </a>
    </div>
    <a
      class="Link-sc-16wd4a0-0 bXBGRT"
      data-gtm-interaction-type="explore"
      data-gtm-trim="GX100"
      title="view details"
    >
      <div
        class="Box-sc-1bsf784-0 jgcAXQ"
        height="139px,248px"
        order="-1"
      >
        <img
          alt=""
          class="sc-1k44dfv-0 jLTFZk"
          draggable="true"
          loading="lazy"
          src="/sample-mock-data/media/img/product-card/engineImage.png"
          style="object-fit: contain;"
        />
      </div>
    </a>
    <div
      class="Box-sc-1bsf784-0 QoaNQ"
    >
      <div
        class="sc-1k44dfv-0 ktMwmo"
      >
        

        <p>
          Day in, day out, get solid performance, reliability and fuel efficiency. Ideal for lawn, construction and agriculture equipment.
        </p>
      </div>
    </div>
    <a
      class="Button-pulw6t-0 sc-1k44dfv-0 eIRfAq"
      data-gtm-interaction-type="explore"
      data-gtm-title="view details"
      data-gtm-trim="GX100"
      data-testid="dsr-button-component"
      display="none,block"
      tabindex="0"
    >
      Shared.Common.viewDetailsButton
    </a>
    <div
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 jkGzMS"
    >
      <label
        class="sc-bwzfXH cvvuMq"
        data-gtm-interaction-type="compare item click"
        data-gtm-title="added comparison item"
        data-gtm-trim="GX100"
        data-testid="product-card-compare-btn"
        data-testvalue="false"
      >
        <input
          aria-checked="false"
          aria-label="Shared.Common.compareCheckboxAria"
          autocomplete="off"
          class="sc-bdVaJa gBQUrE"
          data-gtm-interaction-type="compare item click"
          data-gtm-title="added comparison item"
          data-gtm-trim="GX100"
          id="20ECB7098B9A4584985A7571923A8EB0"
          tabindex="-1"
          type="checkbox"
        />
        <span
          class="sc-htpNat hVaxDi"
          data-testid="cy-compare-toggle"
        />
        <label
          class="Label-sc-115wxjh-0 sc-1k44dfv-0 jLfsOv"
          for="20ECB7098B9A4584985A7571923A8EB0"
        >
          Shared.Common.addToCompareButton
        </label>
      </label>
    </div>
  </div>
</div>
`;
